File: 1463-cherry-pickup-ii/1463-cherry-pickup-ii.py
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
        dp = [[[-1] * n for _ in range(n)] for _ in range(m)]
        dp[0][0][n-1] = grid[0][0] + grid[0][n-1]

        for i in range(1, m):
            for j in range(n):
                for k in range(j+1, n):
                    for x in range(-1, 2):
                        for y in range(-1, 2):
                            if 0 <= j+x < n and 0 <= k+y < n:
                                prev = dp[i-1][j+x][k+y]
                                if prev != -1:
                                    dp[i][j][k] = max(dp[i][j][k], prev + grid[i][j] + (grid[i][k] if j != k else 0))

        ans = max(max(row) for row in dp[m-1])
        return ans if ans != -1 else 0

File: 1463-cherry-pickup-ii/1463-cherry-pickup-ii.cpp
class Solution {
public:
    int dp[70][70][70];
    vector<int>dy = {-1, 0, 1};
   
    int solve(int i, int j1, int j2, vector<vector<int>>& grid){
        if(i == grid.size() - 1){
            return grid[i][j1] + (j1 == j2 ? 0 : grid[i][j2]);
        }
        int &ret = dp[i][j1][j2];
        if(~ret)return ret;
        ret = 0;
        int add = grid[i][j1] + (j1 == j2 ? 0 : grid[i][j2]);
        for(int j = 0; j < 3; ++j){
            int col1 = j1 + dy[j];
            for(int k = 0; k < 3; ++k){
                int col2 = j2 + dy[k];
                if(col1 >= 0 && col1 < grid[0].size() && col2 >= 0 && col2 < grid[0].size()){  
                    ret = max(ret, solve(i + 1, col1, col2, grid) + add);
                }
            }
        }
        return ret;
    }
    int cherryPickup(vector<vector<int>>& grid) {
        memset(dp, -1, sizeof dp);
        return solve(0, 0, grid[0].size() - 1, grid);
    }
};